jfr/src/main/java/io/netty/handler/jfr/JfrEventHandler.java

```diff
@@ -0,0 +1,358 @@
+        event.begin();
+        ctx.fireExceptionCaught(cause);
+        if (event.shouldCommit()) {
+            event.setChannel(ctx.channel());
+            event.commit();
+        }
+    }
+
+    @Override
+    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+        UserEventTriggeredEvent event = new UserEventTriggeredEvent();
+        event.begin();
+        ctx.fireUserEventTriggered(evt);
+        if (event.shouldCommit()) {
+            event.setChannel(ctx.channel());
+            event.userEvent = String.valueOf(evt);
+            event.commit();
+        }
+    }
+
+    @Override
+    public void bind(ChannelHandlerContext ctx, SocketAddress localAddress, ChannelPromise promise) throws Exception {
+        BindEvent event = new BindEvent();
+        event.begin();
+        ctx.bind(localAddress, promise);
+        if (event.shouldCommit()) {
+            event.setChannel(ctx.channel());
+            event.localAddress = String.valueOf(localAddress);
+            event.commit();
+        }
+    }
+
+    @Override
+    public void connect(
+            ChannelHandlerContext ctx,
+            SocketAddress remoteAddress, SocketAddress localAddress, ChannelPromise promise) throws Exception {
+        ConnectEvent event = new ConnectEvent();
+        event.begin();
+        ctx.connect(remoteAddress, localAddress, promise);
+        if (event.shouldCommit()) {
+            event.setChannel(ctx.channel());
+            event.remoteAddress = String.valueOf(remoteAddress);
+            event.localAddress = String.valueOf(localAddress);
+            event.commit();
+        }
+    }
+
+    @Override
+    public void disconnect(ChannelHandlerContext ctx, ChannelPromise promise) throws Exception {
+        DisconnectEvent event = new DisconnectEvent();
+        event.begin();
+        ctx.disconnect(promise);
+        if (event.shouldCommit()) {
+            event.setChannel(ctx.channel());
+            event.commit();
+        }
+    }
+
+    @Override
+    public void close(ChannelHandlerContext ctx, ChannelPromise promise) throws Exception {
+        CloseEvent event = new CloseEvent();
+        event.begin();
+        ctx.close(promise);
+        if (event.shouldCommit()) {
+            event.setChannel(ctx.channel());
+            event.commit();
+        }
+    }
+
+    @Override
+    public void deregister(ChannelHandlerContext ctx, ChannelPromise promise) throws Exception {
+        DeregisterEvent event = new DeregisterEvent();
+        event.begin();
+        ctx.deregister(promise);
+        if (event.shouldCommit()) {
+            event.setChannel(ctx.channel());
+            event.commit();
+        }
+    }
+
+// NOTE: Occurs so often, currently remove for perfomance reasons
+//    @Override
+//    public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
+//        ChannelReadCompleteEvent event = new ChannelReadCompleteEvent();
```

Comment:
What performance issues do you see?

Is it a lot of allocation? If so, there are a few tweaks that can be tried...

If it is not allocation, but many events, you could set a threshold, i.e @Threshold("10 ms)

Reply:
It's an extreme amount of events, which probably messes with allocation as well, I did set a threshold I think, but probably need to raise it.

Reply:
Maybe I am missing something but couldn't we cache the events per Channel somehow and so reduce GC pressure ?

Reply:
That sounds like a good idea. Is the ChannelHandler thread safe? If so then we could just reuse the same event for everything.

Reply:
Yes it is on a per `EventLoop` bases. So all the the `Channel`s that are using the same `EventLoop` will not run on a concurrent basis. That said it may be the easiest to just make it per `Channel` basis. If this is not good enough (and I assume it is) we could also use a `FastThreadLocal`.

Reply:
> It's an extreme amount of events, which probably messes with allocation as well, I did set a threshold I think, but probably need to raise it.

The JIT is usually able eliminate heap allocations, but not always. It depends on how much it must inline. If you can see event objects for the TLAB events, you could cache it, similar to what is done in the JDK. 

Looking up a thread local adds a cost, so should only be done if it is an allocation issue. 

https://github.com/openjdk/jdk/blob/master/src/jdk.jfr/share/classes/jdk/jfr/events/SocketReadEvent.java

Reply:
like I said I think re-using per Channel would be the best.

Reply:
Is there one ChannelHandler per channel if I don't make it Sharable?

Reply:
yes the user will create one per `Channel` and add it to the `ChannelPipeline`.

Reply:
If I were to cache the event per channel, what kind of solution would you recommend @normanmaurer ?

Reply:
just store the event in the ChannelHandler as a field.